### PR TITLE
hal: Improve buffer documentation and cleanup error handling

### PR DESCRIPTION
**Pull request by msiglreith**
Fixes #issue
PR checklist:
- [ ] `make` succeeds (on *nix)
- [ ] `make reftests` succeeds
- [ ] tested examples with the following backends:

### CODE REVIEW

File: src/hal/src/buffer.rs [hunk 1]
```diff
@@ -132,11 +124,10 @@ bitflags!(
...
 
-/// Buffer state
-pub type State = Access;
-
-/// Index buffer view for `bind_index_buffer`, slightly
-/// analogous to an index table into an array.
+/// Index buffer view for `bind_index_buffer`.
+///
+/// Defines a buffer slice used for acquiring the indicies on draw commands.
+/// Indices are used to lookup the vertex indices in the vertex buffers.
```
Comment (kvark) [inline src/hal/src/buffer.rs]: "to lookup vertices" perhaps?

File: src/hal/src/buffer.rs [hunk 2]
```diff
@@ -1,72 +1,58 @@
...
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        let description = self.description();
-        match *self {
-            ViewError::Usage(usage) => write!(f, "{}: {:?}", description, usage),
-            _ => write!(f, "{}", description)
-        }
-    }
+/// Error creating a buffer.
+#[derive(Fail, Debug, Clone, PartialEq, Eq)]
+pub enum BufferCreationError {
```
Comment (kvark) [inline src/hal/src/buffer.rs]: what's the motivation behind having the `Buffer*` prefix for those errors? If we are exposing `buffer` as a module, I find it elegant to keep it `buffer::CreationError`. If we are using the fully qualified names, there is no need to expose the modules.

Comment (msiglreith) [reply src/hal/src/buffer.rs]: I'm not sure which direction we should go for the errors.
Currently having a `error` module containing `DeviceCreationError` and a few more. On the other side we could also move these errors into the different modules as done here with possible name shortening.

Comment (kvark) [reply src/hal/src/buffer.rs]: I'm (non-strongly) in favor of having short error names in respective modules (as opposed to putting them into a dedicated `error` module)

Comment (msiglreith) [reply src/hal/src/buffer.rs]: yeah I think this will be more clear. changed.

**Comment by msiglreith** [comment]
bors r=kvark

**Comment by kvark** [comment]
bors retry

**Comment by msiglreith** [comment]
bors retry

**Comment by msiglreith** [comment]
bors r=kvark

**Comment by kvark** [comment]
bors retry

**Comment by kvark** [comment]
bors retry

**Comment by kvark** [comment]
Bors retry

**Comment by kvark** [comment]
Bors r+

**Pull request merged**